### ISSUE TITLE
Error while downloading large size file over IIS Blazor Server

### ISSUE THREAD

**Issue by merajsyed** [arivera12/BlazorDownloadFile#5]
Getting following error while downloading large size file (around 30 MB) after publishing to IIS

Error: Connection disconnected with error 'Error: WebSocket closed with status code: 1006 ().'.

![Blazor_Large_Download_Error](https://user-images.githubusercontent.com/66589512/84001807-0e536880-a970-11ea-87d2-5d69888784a5.PNG)

**Comment by arivera12** [comment]
@merajsyed it's seems you are getting a timeout from SignalR. 

This is by security design, SignalR max message size by default its only 32 KB.

Take a look of [SignalR Security](https://docs.microsoft.com/en-us/aspnet/core/signalr/security?view=aspnetcore-3.1#buffer-management).

To fix this issues you have options like:

Increase SignalR `ApplicationMaxBufferSize` **_(Not recommended)_**

Download the file from using `HttpClient` then pass the data received data to `BlazorDownloadFile` **_(This should be the way to go)_**

**Comment by merajsyed** [comment]
Same error after increasing `ApplicationMaxBufferSize`. 

I tried by increasing `MaximumReceiveMessageSize` also but no luck.
`services.AddSignalR(e =>
            {
                e.MaximumReceiveMessageSize = 102400000;
             });`

**Comment by arivera12** [comment]
Can you attach a sample which reproduces this error?

**Comment by arivera12** [comment]
Did you increased or tried `TransportMaxBufferSize` as well?

**Comment by merajsyed** [comment]
Yes, I tried by increasing `TransportMaxBufferSize` as well.
I cannot share the sample file due to sensitive data. It is working fine locally. Error is coming only after publishing to IIS.

**Comment by arivera12** [comment]
I haven't try the library over iis and over blazor server so I don't know how and if this works over them, only blazor wasm. 

But you have confirmed that it works over blazor server, that it's good.

I haven't try run blazor over iis so I don't know what could be wrong.

Are you deploying to iis over azure or local iis?

**Comment by arivera12** [comment]
[check out this thread to see if helps](https://stackoverflow.com/questions/60148492/client-cannot-connect-to-signalr-hub-of-a-blazor-webassembly-app-running-on-iis/60154089#60154089)

**Comment by arivera12** [comment]
The only way to fix this may be comparing you iis express setting vs your iis settings in production envinroment.

**Comment by merajsyed** [comment]
> I haven't try the library over iis and over blazor server so I don't know how and if this works over them, only blazor wasm.
> 
> But you have confirmed that it works over blazor server, that it's good.
> 
> I haven't try run blazor over iis so I don't know what could be wrong.
> 
> Are you deploying to iis over azure or local iis?

**Comment by merajsyed** [comment]
Accidentally closed the issue.
I am deploying it over local iis and have web socket enabled.
It seems other people also have same issue reported with aspnetcore [https://github.com/dotnet/aspnetcore/issues/21546](https://github.com/dotnet/aspnetcore/issues/21546)

**Comment by arivera12** [comment]
The link shared redirects to this repository.

**Comment by arivera12** [comment]
I copy the text instead.

**Comment by arivera12** [comment]
I read the issue, its seems a problem with blazor server with deployment host, [related](https://github.com/dotnet/aspnetcore/issues/21546). Have you tried or have a chance to deploy this on top of blazor wasm? If not possible, I just ran out of ideas, we would need to wait for a fix.

**Comment by merajsyed** [comment]
I ended up downloading in chunks as used in this [gist](https://gist.github.com/JPVenson/8eb2686b2f07d28014d0f4098d0e04c4) and it worked.

**Comment by arivera12** [comment]
@merajsyed so It's seems I may have a bug with max capacity of downloading a file... I will keep this issue open to upgrade the library with buffer management. Thanks for the link.

**Comment by Smurf-IV** [comment]
Can "the buffer management" solution, also have a way of 
- Starting a save
- Adding many buffers (as they are downloaded to save having to create a single monster ManyGB memory stream and blowing the blazor VM)
- Committing the save (Or auto dispose if exception thrown etc.)

As per a rework of the gist above ?

**Comment by arivera12** [comment]
@Smurf-IV I transfered your comment into a new issue since sounds promising and a bit complicated because I also found some blazor issues and downsides on stackoverflow.

See the transfered issue here  https://github.com/arivera12/BlazorDownloadFile/issues/7

**Comment by arivera12** [comment]
I almost got this. But have been struggling for some hours to fix byte[] and Stream overload methods when converting to base64 for some reason atob works for the first buffer but not for the rest of the buffers in my test example. Base64 string buffered method work perfectly. I may take a deeper look on the weekend. If anyone wants to take a look and help me, this is the commit https://github.com/arivera12/BlazorDownloadFile/commit/f64ef3db54b1d11d36778d61b633d74ada972938

**Comment by arivera12** [comment]
Ok I just figured out what was wrong, it won't work after decoding cause the root of the base64 string is need for decoding correctly. I will try another approach I have in mind.
